component/amqp/component.go

```diff
@@ -0,0 +1,359 @@
+// Package amqp provides a native consumer for the AMQP protocol.
+package amqp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/beatlabs/patron/correlation"
+	patronerrors "github.com/beatlabs/patron/errors"
+	"github.com/beatlabs/patron/log"
+	"github.com/beatlabs/patron/trace"
+	"github.com/google/uuid"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/streadway/amqp"
+)
+
+type messageState string
+
+const (
+	defaultBatchCount        = 1
+	defaultBatchTimeout      = 1<<63 - 1 // max time duration possible effectively disabling the timeout.
+	defaultHeartbeat         = 10 * time.Second
+	defaultConnectionTimeout = 30 * time.Second
+	defaultLocale            = "en_US"
+	defaultStatsInterval     = 5 * time.Second
+	defaultRetryCount        = 10
+	defaultRetryDelay        = 5 * time.Second
+
+	consumerComponent = "amqp-consumer"
+
+	ackMessageState     messageState = "ACK"
+	nackMessageState    messageState = "NACK"
+	fetchedMessageState messageState = "FETCHED"
+)
+
+var (
+	messageAge     *prometheus.GaugeVec
+	messageCounter *prometheus.CounterVec
+	queueSize      *prometheus.GaugeVec
+)
+
+func init() {
+	messageAge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "amqp",
+			Name:      "message_age",
+			Help:      "Message age based on the AMQP timestamp",
+		},
+		[]string{"queue"},
+	)
+	prometheus.MustRegister(messageAge)
+	messageCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: "amqp",
+			Name:      "message_counter",
+			Help:      "Message counter by state and error",
+		},
+		[]string{"queue", "state", "hasError"},
+	)
+	prometheus.MustRegister(messageCounter)
+	queueSize = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "amqp",
+			Name:      "queue_size",
+			Help:      "Queue size reported by AMQP",
+		},
+		[]string{"queue"},
+	)
+	prometheus.MustRegister(queueSize)
+}
+
+// ProcessorFunc definition of a async processor.
+type ProcessorFunc func(Batch)
+
+type batchConfig struct {
+	count   uint
+	timeout time.Duration
+}
+
+type retryConfig struct {
+	count uint
+	delay time.Duration
+}
+
+type statsConfig struct {
+	interval time.Duration
+}
+
+// Component implementation of a async component.
+type Component struct {
+	url      string
+	queue    string
+	requeue  bool
+	proc     ProcessorFunc
+	batchCfg batchConfig
+	statsCfg statsConfig
+	retryCfg retryConfig
+	cfg      amqp.Config
+	traceTag opentracing.Tag
+	mu       sync.Mutex
```

Comment:
i am wondering if it maybe make sense to use a pointer for the mutex ?

Reply:
Funny enough, the mutex is actually unnecessary!